### PR TITLE
add new testing system, merge master on union types branch

### PR DESCRIPTION
**Pull request by osdiab**
It wasn't clear what your change was outputting due to the lack of actual tests, so I merged master and made the tests pass based on what your code does.

### CODE REVIEW

File: tests/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        oneOf: iots.keyof({ a: null, b: null, c: null }),
+        stringUnion: iots.union([iots.literal('a'), iots.literal('b'), iots.literal('c')]),
+        interfaceUnion: iots.union([
+            iots.interface({ key: iots.string }),
+            iots.interface({ code: iots.number }),
+        ]),
+    });
+
+    t.deepEqual(reporter(Unions.decode({
+        oneOf: '',
```
Comment (osdiab) [inline tests/index.test.ts:78]: it appears that as is, `keyof` is not outputting an error properly given that this passes

File: tests/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                gender: Gender,
+            })
+        )
+    });
+    const result = Person.decode({
+        name: 'Giulio',
+        children: [{ gender: 'Whatever' }]
+    });
+
+    t.deepEqual(reporter(result), [
```
Comment (osdiab) [inline tests/index.test.ts:55]: this test's output changed as well as a result of your change

File: tests/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+});
+
+test('formats nested array item mismatches correctly', (t) => {
+    const NumberGroups = iots.array(iots.array(iots.number));
+    const result = NumberGroups.decode([[{}]]);
+
+    t.deepEqual(reporter(result), ['Expecting number at 0.0 but instead got: {}.']);
+    t.end();
+});
+
```
Comment (osdiab) [inline tests/index.test.ts:37]: everything above remains unchanged from master

**Comment by gillchristian** [comment]
Thanks for this!

> so I merged master and made the tests pass based on what your code does

We should actually have what is the expected output as tests. I'll do that on my PR.

**Pull request merged**